aten/src/ATen/Context.cpp

```diff
@@ -263,7 +263,20 @@ void Context::setLinalgPreferredBackend(at::LinalgBackend b) {
   }
 }
 
-at::BlasBackend Context::blasPreferredBackend() const {
+at::BlasBackend Context::blasPreferredBackend() {
+#ifdef USE_ROCM
+  if (blas_preferred_backend == at::BlasBackend::Cublaslt) {
+    static const std::vector<std::string> archs = {"gfx90a", "gfx940", "gfx941", "gfx942"};
```

Comment:
Having yet another place where an arbitrary list of llvm targets is placed seams like a bad idea, since it will have to be remembered when the supported targets of hipblaslt expand or contracts. Further this list is already wrong right now as hipblaslt has support for some gfx11 targets.

At the very least this needs to be a define set via a cmake option, but ideally you would query the architectures from the hipblaslt binary which is not that hard to implement directly, but ideally the runtime would of course provide this infomation

Reply:
You make a good point about the maintenance headache this introduces. I'm not sure about this being a cmake option though, since this is not exactly user-configurable information? I'm looking into whether the hipblasLT library provides us a way to query the list of supported archs.

Reply:
So hipblasLT doesn't currently have an API to report supported gfx archs, but we will request that. Until then, I believe this solution is appropriate.

Reply:
I think that's fair.

Id like to also mention here that this code currently dose not work at all at the moment since haveing one of the gpus in the system that are not supported by hipblaslt causes the runtime to assert here when a hipblaslt code object is loaded by ldd: https://github.com/ROCm/clr/blob/204d35d16ef5c2c1ea1a4bb25442908a306c857a/hipamd/src/hip_code_object.cpp#L762 from https://github.com/ROCm/clr/blob/204d35d16ef5c2c1ea1a4bb25442908a306c857a/hipamd/src/hip_code_object.cpp#L752C22-L752C30 which ultimately calls ExtractFatBinaryUsingCOMGR

In the tests on ci this appears to work as you have disabled runtime assertions in clr there (which is imo not great in and of itself), but it dosent really work. When you do have a supported and an unsupported gpu in the system, depending on the gpu order ExtractFatBinaryUsingCOMGR can fail and return before it gets to the supported gpu, this causes the gpu code objects to subsequently be missing when torch tries to use them.

I presume a solution for this is in the pipe, because at the moment the way this pr attempts to select which gpus to use hipblaslt on at runtime simply dose not work with the rocm runtime as by the time the above code is run the runtime has already entered a failed state.

Reply:
@IMbackK I agree that there's an issue in the way HIP runtime handles code object loading for multiple GPUs in a heterogenous system. However, this PR actually intends to set the blas_backend to `at::BlasBackend::Cublas` if *any* of the GPUs in the system are unsupported. This means that if you have a system with a gfx90a and a gfx908 GPU, trying to set the preferred backend to `at::BlasBackend::Cublaslt` will end up overriding it to `at::BlasBackend::Cublas`. IIUC, that should not break functionality. In other words, this PR is *not* attempting to "select which gpus to use hipblaslt on at runtime", it is either using hipblasLT on all GPUs (if they're all supported), or on none of them.
If you do have a heterogenous system, please try this PR on it and confirm if you observe the above behaviour.

Reply:
Well the issue is that on a gfx906/gfx908/gfx10xx (only, non heterogeneous) system we never even get here at all, merely linking hipblaslt causes clr to assert before ldd even finishes loading libtorch_hip.so unless one disables assertions in clr, which 1. is just sweeping a real problem under the rug and 2. is unacceptable on a dev system since it hides other problems.

Reply:
@IMbackK Please correct me if I'm wrong, but this is how I understand the concerns you are raising:
* There's a HIP runtime issue which causes a functional issue on unsupported gfx archs (on hetero or homogenous systems) _only when runtime assertions are enabled in clr_
* This PR is to prevent users from setting the wrong/unsupported cublaslt backend if *any* of their GPUs do not support it (_regardless_ of the clr assert issue)
* Setting the default value of the preferred linalg backend to cublas will _still_ run into the clr assertions since PyTorch will still try to load hipblaslt library via libtorch_hip.so at the start

Can you please confirm the above matches your understanding?

Reply:
yes all those points are correct. 

the only slight nitpick i have is that I dont know if the problem in the HIP runtime is an issue per-say, the runtime simply dosent support loading objects containing hip code but not containing code objects for all available gpus and the clr code is pretty explicit about that attempting this is in fact an error. I guess it is more a missing feature.

Reply:
Do you have a link to an issue that has been filed for the assertion-enabled scenario? I think we should follow-up on that to see how we can resolve it properly.

Reply:
https://github.com/pytorch/pytorch/issues/119081 is is issue in a way and i am currently using https://github.com/pytorch/pytorch/pull/120551 however https://github.com/pytorch/pytorch/pull/120551 can really only be considered a solution unless pytorch disables hipblaslt at compile time using that pr for all official builds that are supposed to support gpus besides CDNA2/3 and RDNA3 until one of the following happens:
1. hipblaslt changes to not have gpu code in the main .so but to instead load all gpu code as hipmodules
2. hipblaslt gains support for all the usual rocm targets
3. the runtime gains support for loading code objects that lack support for a given gpu and gains api to determine when this has occurred so that clients can avoid calling into these code objects.